src/background/mod.rs

```diff
@@ -41,18 +88,234 @@ pub struct PlanetComponent {
     for (mut transform, planet) in query.iter_mut() {
-        transform.rotation *= Quat::from_rotation_y(planet.rotation_speed);
+        transform.rotation *= Quat::from_rotation_y(planet.rotation_speed * time.delta_seconds());
     }
 }
 
-/// Create background from resource
+/// Execute the exploding star effect if the game is lost through defense being destroyed
+pub fn on_defeat_star_explode_system(
+    mut run_end_event_reader: EventReader<RunEndEvent>,
+    mut point_light_query: Query<&mut PointLight, With<StarLightComponent>>,
+    mut star_explode_res: ResMut<StarExplodeResource>,
+    time: Res<Time>,
+    backgrounds_res: Res<BackgroundsResource>,
+) {
+    // Check for loss condition from defense objective
+    for event in run_end_event_reader.iter() {
+        if let RunOutcomeType::Defeat(RunDefeatType::DefenseDestroyed) = event.outcome {
+            star_explode_res.started = true;
+        }
+    }
+
+    // Update star point light intensity if star explosion active
+    if star_explode_res.started {
+        for mut point_light in point_light_query.iter_mut() {
+            point_light.intensity *= backgrounds_res.star_explode_intensity * time.delta_seconds();
+        }
+    }
+}
+
+/// Create a procedurally generated 3D background for a level
 pub fn create_background_system(
     mut commands: Commands,
     asset_server: Res<AssetServer>,
-    backgrounds: Res<BackgroundsResource>,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<StandardMaterial>>,
+    mut star_explode_res: ResMut<StarExplodeResource>,
+    backgrounds_res: Res<BackgroundsResource>,
 ) {
-    backgrounds.backgrounds["solar_system"].spawn(&mut commands, &asset_server);
+    // reset the star explode reource
+    *star_explode_res = StarExplodeResource::default();
+
+    let mut rng = rand::thread_rng();
+
+    // Choose random positions for the bodies
+    let background_transform = Transform::from_translation(backgrounds_res.background_transation);
+    let star_transform = Transform::from_xyz(
+        rng.gen_range(backgrounds_res.star_position_x_range.clone()),
+        0.0,
+        rng.gen_range(backgrounds_res.star_position_z_range.clone()),
+    );
+    let planet_transform = Transform::from_translation(backgrounds_res.planet_translation);
+
+    // Choose a random planet or fallback to a white icosphere
+    {
+        let mut planet_commands = commands.spawn_empty();
+        planet_commands
+            .insert(PlanetComponent {
+                rotation_speed: rng.gen_range(backgrounds_res.rotation_speed_range.clone()),
+            })
+            .insert(GameCleanup)
+            .insert(Visibility::default())
+            .insert(ComputedVisibility::default())
+            .insert(Name::new("Planet"));
+
+        match fs::read_dir("./assets/models/planets")
+            .expect("Failed to access ./assets/models/planets")
+            .choose(&mut rng)
```

Comment:
Maybe hoist file IO to another function so that we can use the `?` operator for more error handling and have 1 place to fall back when any of it fails.